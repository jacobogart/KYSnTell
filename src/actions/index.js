export const setLocations = (locations) => ({
  type: 'SET_LOCATIONS',
  locations
})