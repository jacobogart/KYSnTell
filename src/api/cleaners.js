export const locationCleaner = (location) => {
  const { title, link, id, streetAddress, locality, region, postalCode, telephone, point} = location;
  const shortId = id.split('=')[1];
  const address = `${streetAddress};${locality}, ${region} ${postalCode}`;
  return {
    title,
    link,
    id: shortId,
    address,
    telephone,
    point
  }
}