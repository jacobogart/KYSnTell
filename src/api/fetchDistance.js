import { googleKey, googleRoot } from "./utilities";

export const fetchDistance = (origin, destination) => {  return fetch(`${googleRoot}distancematrix/json?origins=${origin.lat},${origin.lng}&destinations=${destination.lat},${destination.long}&units=imperial&key=${googleKey}`)
    .then(response => {
      if(!response.ok) {
        throw Error('Failed to find distance')
      } else {
        return response.json()
      }
    })
    .then(result => result.rows[0].elements[0].distance.text)
}