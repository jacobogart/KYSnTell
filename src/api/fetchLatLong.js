import { geocodeUrl } from "./utilities";

export const fetchLatLong = (zip) => {
  const googleKey = process.env.REACT_APP_GOOGLE_KEY;
  return fetch(`${geocodeUrl}${zip}&key=${googleKey}`)
    .then(response => {
      if(!response.ok) {
        throw Error('Failed to find coordinates')
      } else {
        return response.json()
      }
    })
    .then(result => result.results[0].geometry.location)
}