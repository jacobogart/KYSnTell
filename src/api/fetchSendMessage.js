export const fetchSendMessage = (contacts, message) => {
  const messagePromises = contacts.map(contact => {
      const body = { to: contact, message}
      const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(body)
    }
    return fetch('/api/messages', options)
      .then(response => {
        if (!response.ok) {
          throw Error('Failed to send message')
        } else {
          return response.json();
        }
      })
    })
  return Promise.all(messagePromises)
}
  