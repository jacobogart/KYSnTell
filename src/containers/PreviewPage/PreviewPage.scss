.message-holder {
  background-color: grey;
  width: 85vw;
  padding: 10px;

  .message {
    color: black;
    font-size: 18px;
  }
}

.PreviewPage {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  height: 70vh;
  margin-top: 15vh;

  button {
    background-color: white;
    font-size: 20px;
    padding: 8px;
    display: block;
    width: 90vw;
    border-radius: 10px;
  }
}