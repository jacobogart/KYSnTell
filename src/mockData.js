import { locationCleaner } from "./api/cleaners";

export const origin = {
  lat: 38.9119463, 
  lng: -94.4136213 
}

export const destination = { 
  lat: "39.1952922", 
  lng: "-94.6817269" 
}

export const distance = '18.3 mi';

export const distanceResult = {
  rows: [
    {
      elements: [
        {
          distance: {
            text: distance
          }
        }
      ]
    }
  ]
}


export const fetchDistanceURL = "https://cors-anywhere.herokuapp.com/https://maps.googleapis.com/maps/api/distancematrix/json?origins=38.9119463,-94.4136213&destinations=39.1952922,-94.6817269&units=imperial&key=AIzaSyDHKnndIp7I8oymrxO0sDwpQUdZvfN1sqc"

export const latLongResult = {
  results: [
    {
      geometry: {
        location: origin
      }
    }
  ]
}


export const fetchLatLongURL = "https://maps.googleapis.com/maps/api/geocode/json?address=64081&key=AIzaSyDHKnndIp7I8oymrxO0sDwpQUdZvfN1sqc"

export const locations = [
  {
    title: "Neighborhood Health",
    link: "https://gettested.cdc.gov/gettested_redirect/21952",
    id: "tag:hivtest.org,2000-03-20:/search/Detailx.asp?id=21952",
    streetAddress: "617 S 8th St",
    locality: "Nashville",
    region: "TN",
    postalCode: "37206-3819",
    telephone: "(615)-227-3000",
    categories: "",
    point: {
      lat: "36.1678930",
      long: "-86.7547047"
    },
    updated: "2019-02-08T17:19:32.0000000-05:00"
  },
  {
    title: "Street Works",
    link: "https://gettested.cdc.gov/gettested_redirect/110792",
    id: "tag:hivtest.org,2008-05-21:/search/Detailx.asp?id=110792",
    streetAddress: "520 Sylvan St",
    locality: "Nashville",
    region: "TN",
    postalCode: "37206",
    telephone: "(615)-259-7676",
    categories: "",
    point: {
      lat: "36.1665114",
      long: "-86.7605114"
    },
    updated: "2018-08-02T18:05:30.0000000-04:00"
  }
];

export const locationsResult = {
  services: [
    {
      providers: locations
    }
  ]
}

export const cleanLocations = locations.map(location => locationCleaner(location))

export const fetchLocationsURL = "https://cors-anywhere.herokuapp.com/https://locator.hiv.gov/data?lat=38.9119463&long=-94.4136213&distance=10"

export const contact = '+18168682142';

export const contacts = [contact, contact]

export const message = 'Test message.'

export const sendMessageOptions = {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json'
  },
  body: JSON.stringify({ to: contact, message })
}

export const resolvedMessages = [{ ok: true }, { ok: true }];

export const cleanLocation = {
  title: 'Neighborhood Health',
  link: 'https://gettested.cdc.gov/gettested_redirect/21952',
  id: '21952',
  address: '617 S 8th St;Nashville, TN 37206-3819',
  telephone: '(615)-227-3000',
  point: { lat: '36.1678930', lng: '-86.7547047' }
}

export const details = {
  additionalNotes: "Test",
  diagnosis: "Herpes",
  timeFrame: "year"
}

export const user = {
  location: origin
}

export const searchChangeEvent = {
  target: {
    name: 'zipcode',
    value: '64081'
  }
}

export const submitEvent = {
  preventDefault: jest.fn()
}

export const history = {
  push: jest.fn()
}

export const setUserLocation = jest.fn();

export const setLocations = jest.fn();

export const setContacts = jest.fn();

export const fullMessage = 'Hello, we have been informed by an anonymous sexual partner that you may have been exposed to Herpes in the last year. While this is no cause for alarm, we do recommend getting tested at your earliest convenience. To find a testing center near you, please visit kysntell.com, or contact your preferred healthcare provider. Additional notes from partner: Test'

export const messageWONotes = "Hello, we have been informed by an anonymous sexual partner that you may have been exposed to Herpes in the last year. While this is no cause for alarm, we do recommend getting tested at your earliest convenience. To find a testing center near you, please visit kysntell.com, or contact your preferred healthcare provider."

export const previewEvent = {
  target: {
    blur: jest.fn()
  }
}

export const messageContext = {
  focus: jest.fn(),
  blur: jest.fn(),
  innerText: 'Test'
}

 export const rejectedMessages = [{ ok: false, contact }]

 export const routerLocation = {
  pathname:'/kys/locations/21952'
 }

export const detailsChangeEvent = {
  target: {
    name: 'diagnosis',
    value: 'Herpes'
  }
}

export const setDetails = jest.fn();

export const contactChangeEvent = {
  target: {
    value: contact
  }
}

export const contactBadChangeEvent = {
  target: {
    value: 'test'
  }
}

export const searchDefaultState = {
  zipcode: '',
  distance: '',
  isLoading: false,
  error: false
}