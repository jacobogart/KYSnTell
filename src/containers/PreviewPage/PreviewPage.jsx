import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchSendMessage } from '../../api/fetchSendMessage'
import { setContacts } from '../../actions';


export class PreviewPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      message: '',
      contEdit: false
    }
    this.message = React.createRef();
  }

  componentDidMount() {
    this.setMessage();
  }

  setMessage = () => {
    const { diagnosis, timeFrame, additionalNotes } = this.props.details;
    let message = `Hello, we have been informed by an anonymous sexual partner that you may have been exposed to ${diagnosis} in the last ${timeFrame}. While this is no cause for alarm, we do recommend getting tested at your earliest convenience. To find a testing center near you, please visit kysntell.com, or contact your preferred healthcare provider.`
    if (additionalNotes) { message = message.concat(' ', `Additional notes from partner: ${additionalNotes}`) }
    this.setState({ message });
  }
  
  toggleEdit = (e) => {
    const { contEdit } = this.state;
    if (contEdit) {
      this.setState({ message: this.message.current.innerText });
      this.message.current.blur()
    } else {
      e.target.blur();
      this.message.current.focus();
    }
    this.setState({ contEdit: !contEdit });
  }

  handleSubmit = () => {
    fetchSendMessage(this.props.contacts, this.state.message)
      .then(res => {
        if (res.every(res => res.ok)) {
          this.props.history.push('/tell/success')
        } else {
          const failedContacts = res.filter(res => !res.ok).map(res => res.contact);
          console.log(failedContacts);
          this.props.setContacts(failedContacts);
          this.props.history.push('/tell/error')
        }
      })
  }

  render() {
    const { contEdit, message } = this.state;
    const buttonText = contEdit
      ? 'Save Changes'
      : 'Edit Message'
    return (
      <div className="PreviewPage">
        <div className="message-holder">
          <p
            tabIndex="0"
            ref={this.message}
            contentEditable={contEdit} 
            className={`message edit-${contEdit}`}
          >
            {message}
          </p>
        </div>
        <button onClick={this.toggleEdit}>{buttonText}</button>
        <button onClick={this.handleSubmit}>Send Message</button>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  contacts: state.contacts,
  details: state.details
});

export const mapDispatchToProps = (dispatch) => ({
  setContacts: (contacts) => dispatch(setContacts(contacts))
})

export default connect(mapStateToProps, mapDispatchToProps)(PreviewPage);