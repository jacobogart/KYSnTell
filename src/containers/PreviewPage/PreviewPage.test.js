import React from 'react';
import { shallow } from 'enzyme';
import { PreviewPage, mapStateToProps, mapDispatchToProps } from './PreviewPage';
import * as mock from '../../mockData'
import { fetchSendMessage } from '../../api/fetchSendMessage';
import { setContacts } from '../../actions';

jest.mock('../../api/fetchSendMessage')

describe('PreviewPage', () => {
  let wrapper, instance;

  fetchSendMessage.mockImplementation(() => Promise.resolve(mock.resolvedMessages))

  beforeEach(() => {
    wrapper = shallow(
      <PreviewPage 
        details={mock.details}
        contacts={mock.contacts}
        history={mock.history}
        setContacts={mock.setContacts}
      />);
    instance = wrapper.instance();
    instance.message.current = mock.messageContext;
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    wrapper = shallow(
      <PreviewPage
        details={mock.details}
        contacts={mock.contacts}
      />, { disableLifecycleMethods: true });
    const defaultState = {
      message: '',
      contEdit: false
    };

    expect(wrapper.state()).toEqual(defaultState);
  });

  it('should invoke setMessage on mount', () => {
    jest.spyOn(instance, 'setMessage');
    instance.componentDidMount();
    expect(instance.setMessage).toHaveBeenCalled();
  });

  describe('Event Listeners', () => {
    it('should invoke toggleEdit when edit-btn is clicked', () => {
      expect(wrapper.state('contEdit')).toEqual(false);
      wrapper.find('.edit-btn').simulate('click', mock.previewEvent);
      expect(wrapper.state('contEdit')).toEqual(true);
    });

    it('should invoke handleSend when send-btn is clicked', () => {
      wrapper.find('.send-btn').simulate('click');
      expect(fetchSendMessage).toHaveBeenCalled();
    });
  });

  describe('setMessage', () => {
    it('should set state to the message', () => {
      instance.setMessage();
      expect(wrapper.state('message')).toEqual(mock.fullMessage);
    });

    it('should not include additionalNotes if there are none', () => {
      wrapper = shallow(
        <PreviewPage
          details={ {...mock.details, additionalNotes: ''}}
          contacts={mock.contacts}
        />, { context: mock.messageContext});
      expect(wrapper.state('message')).toEqual(mock.messageWONotes);
    });
  });

  describe('toggleEdit', () => {
    it('should invoke blur on the target element if contEdit is false', () => {
      instance.toggleEdit(mock.previewEvent);
      expect(mock.previewEvent.target.blur).toHaveBeenCalled();
    });

    it('should invoke focus on the message if contEdit is false', () => {
      instance.toggleEdit(mock.previewEvent);
      expect(instance.message.current.focus).toHaveBeenCalled();
    });

    it('should set state to the update message if contEdit is true', () => {
      instance.setState({ contEdit: true });
      instance.toggleEdit(mock.previewEvent);
      expect(wrapper.state('message')).toEqual('Test');
    });

    it('should blur the message if contEdit is true', () => {
      instance.setState({ contEdit: true });
      instance.toggleEdit(mock.previewEvent);
      expect(instance.message.current.blur).toHaveBeenCalled();
    });

    it('should toggle the contEdit piece of state', () => {
      expect(wrapper.state('contEdit')).toEqual(false);
      instance.toggleEdit(mock.previewEvent);
      expect(wrapper.state('contEdit')).toEqual(true);
      instance.toggleEdit(mock.previewEvent);
      expect(wrapper.state('contEdit')).toEqual(false);
    });
  });

  describe('handleSend', () => {
    it('should call fetchSendMessage', () => {
      instance.handleSend();
      expect(fetchSendMessage).toHaveBeenCalledWith(mock.contacts, mock.fullMessage);
    });

    it('should push the router if all responses are ok', async () => {
      await instance.handleSend();
      expect(mock.history.push).toHaveBeenCalledWith('/tell/success');
    });

    it('should invoke handleError is there are any failed fetches', async () => {
      fetchSendMessage.mockImplementation(() => Promise.resolve([...mock.resolvedMessages, { ok: false, contact: mock.contact}]));
      jest.spyOn(instance, 'handleError');
      await instance.handleSend();
      expect(instance.handleError).toHaveBeenCalled();
    });
  });

  describe('handleError', () => {
    it('should invoke setContacts', () => {
      instance.handleError(mock.rejectedMessages);
      expect(mock.setContacts).toHaveBeenCalledWith([mock.contact])
    });

    it('should push the router to the error page', () => {
      instance.handleError(mock.rejectedMessages);
      expect(mock.history.push).toHaveBeenCalledWith('/tell/error');
    });
  });

  describe('mapStateToProps', () => {
    it('should return the correct pieces of state', () => {
      const { contacts, details, user } = mock;
      const mockState = { contacts, details, user };
      const expected = { contacts, details }
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    });
  });

  describe('mockDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const actionToDispatch = setContacts(mock.contacts);
    mappedProps.setContacts(mock.contacts);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});