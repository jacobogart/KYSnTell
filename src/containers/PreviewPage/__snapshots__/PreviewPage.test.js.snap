// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewPage should match snapshot 1`] = `
<div
  className="PreviewPage"
>
  <div
    className="message-holder"
  >
    <p
      className="message edit-false"
      contentEditable={false}
      tabIndex="0"
    >
      Hello, we have been informed by an anonymous sexual partner that you may have been exposed to Herpes in the last year. While this is no cause for alarm, we do recommend getting tested at your earliest convenience. To find a testing center near you, please visit kysntell.com, or contact your preferred healthcare provider. Additional notes from partner: Test
    </p>
  </div>
  <button
    className="edit-btn"
    onClick={[Function]}
  >
    Edit Message
  </button>
  <button
    className="send-btn"
    onClick={[Function]}
  >
    Send Message
  </button>
</div>
`;
